### DISCUSSION TITLE
mouseDown on uiImageButton

### DISCUSSION THREAD

**Discussion by bijlevel** [RomanKubiak/ctrlr#587] [Q&A] [answered]
I run into the (already known) problem of a uiImageButton not recognizing mouseDown (or Up for that matter). I did found a discussion on ctrlr.org about this. It turns out that a uiImageButton does only react on a mouseclick when one clicks on a label. But the image doesn't change!  I think I can overcome this problem, by making a script that is called when  the mouse is down on the label and then in that script getting the value of the button and change the button value to 0 or 1, depending on the initial value. Would that work? Also, since we have now a few VERY experienced coders ;-) (Damian and Ted-juh), is it possible to change the source, because it looks like a bug in CTRLR?

**Comment by bijlevel** [comment]
I did experiment further with this "issue". I can call the needed script when the button value changes, but at the risk (I found out) of having an infinite loop, because my script calls a midireceived script which causes the button value to change and then this button script is called again etc. etc. ;-). So I will study the discussion about this on the ctrlr.org, because I remember seeing same "loop" issues, which were solved at the end, I believe. Still I whished it could be solved by altering the source code of CTRLR...

**Comment by dobo365** [comment]
Hi bijlevel!
I don't have any issue with that on my panels and I'm only using uiImageButtons.
No loop issues either.
I have an "_OnClick" method assigned to Called when a mouse is down property
I have an "_OnChange" method assigned to Called when the modulator value changes
The OnClick method is used to display some info text
The OnChange is calling the OnClick method
When you click on the label, only the info text is shown without changing the button
When you click on the button, the value and image are changed and the info text is also shown

**Let me know if you need more details.**

Thx for your work on the Ctrlr build for M1. I'm chasing a second hand M1 now just because of you. Hahaha...
On my side, I started to play and understand the way the properties are coded and displayed.
As you say and as Damien is saying, we are all looking for a different piece of code which is good.
Sometimes we can bring all our knowledge together and make some good documentation of the code.
I'm doing that already in parallel of my findings.

Happy New Year....Dominique

**Reply by bijlevel** [reply]
OK, I will study those scripts in your panel(s). I think indeed it is just a matter of proper Lua scripting.

Glad I could be a bit of a deciding factor in your computer purchase. BTW the intel version of CTRLR does also run fine on my M2.

And yes I like this github place very much because of the good spirit amongst the contributors and the willingness to help each other out!

Happy paneling for 2024! Sjoerd

**Reply by dobo365** [reply]
The OnChange method is like this:

```
--
-- Called when a modulator value changes
-- @mod   http://ctrlr.org/api/class_ctrlr_modulator.html
-- @value    new numeric value of the modulator
--
OnOff_OnChange = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

	-- No action while loading a program or if the panel is in bootstrap or program states
	if bLoadingProgram or not isPanelReady() then
		return
	end

	OnOff_OnClick(mod:getComponent())

end
```

The OnClick method is like this:

```
--
-- Called when a mouse is down on this component
--

OnOff_OnClick = function(--[[ CtrlrComponent --]] comp, --[[ MouseEvent --]] event)

	-- No action while the panel is in bootstrap state
	if not isPanelReady() or not bLCD then
		return
	end

	local sName= L(comp:getOwner():getName())
	local bGlobal = false

	-- Get parameter name to be displayed
	if sName == "PortamentoMode" then 
		sParameter = "Portamento Mode (Auto, On)"
		sDescription = "Auto: When playing in a legato style (playing a new note before any previously played note is released), portamento will be applied.\nOn: Portamento will be applied, even when fully releasing a key and then playing the next note."
		sON = "On"
		sOFF = "Auto"
		modMain = modPortamentoMode
		if LoadedProgramData~=nil then
			iSavedMain = LoadedProgramData:getByte(133)
		else
			iSavedMain = nil
 		end
	elseif sName == "PortamentoBPMSync" then 
		sParameter = "Portamento BPM Sync (Off, On)"
		sDescription = "When the BPM Sync is set to ON, the length of portamento that is set using the PORTAMENTO knob will be synchronized with the BPM (beats per minute)."
		sON = "On"
		sOFF = "Off"
		modMain = modPortamentoBPMSync
		if LoadedProgramData~=nil then
			iSavedMain = LoadedProgramData:getByte(134)
		else
			iSavedMain = nil
		end
...
	end

	-- Display current and saved values for ON/OFF (up/down)
	if not bLoadingProgram then
		txtLCD_Parameter:getComponent():setPropertyString ("uiLabelText", sParameter)
		txtLCD_Description:getComponent():setPropertyString ("uiLabelText", sDescription)
...
		if modMain ~= nil then
			if modMain:getValue()==1 then
				txtLCD_CurrentValue:getComponent():setPropertyString ("uiLabelText", sON)
			else
				txtLCD_CurrentValue:getComponent():setPropertyString ("uiLabelText", sOFF)
			end
			if iSavedMain~=nil then
				if iSavedMain==1 then
					txtLCD_SavedValue:getComponent():setPropertyString ("uiLabelText", sON)
				elseif iSavedMain == 0 then
					txtLCD_SavedValue:getComponent():setPropertyString ("uiLabelText", sOFF)
				end
			else
				txtLCD_SavedValue:getComponent():setPropertyString ("uiLabelText", "nil")
			end
		end
	end

end
```

**Reply by bijlevel** [reply]
Thanks! ATM I'm fine with a call when the value of the button changes, That's not an problem anymore.

I'm now struggling with the BigInteger operations. In my panel I use getBitRangeAsInt to get (the value of)  a bit at a certain position in the byte. But I need to change the value of that bit. So I thought I could use setBitRangeAsInt, but this does not work, because I probably use the wrong syntax. This is what I have sofar:

__flags = BigInteger(db)_ -- db is the "byte" (with a value e.g. of 4) for which I need to change the 4th bit
_bit1 = flags.getBitRangeAsInt(3,1)_ -- gets the bit value of the 4th bit_

This gives me the value of bit1, in this case 0 (the number 4 in bits is 0000 0100). Then I want to change the value of db by inserting a new value for bit1, so the result is 0000 1100, which is decimal number 12 . So I first change bit1 to 1 and try to set the bit range like this:

_bit1 =1
db = flags.setBitRangeAsInt(3,1,bit1)_

But now Lua tells me that **db** is a nil value...

 I hope the problem is clear. So what am I doing wrong?

**Reply by dobo365** [reply]
For that kind of thing you should just use bitwise operations (right shifts, and/or). 
Look at that in the Lua documentation then experiment a bit.
To know the value of bit 4, you right shift your byte by 3 so that bit 4 is in the first position then you do an "and" with 1
`	modStep4OnOff:setValue(bit.band(bit.rshift(LoadedProgramData:getByte(170),3),1), true)`
To set the value of bit 4, you take your bit value, do a left shift by 3 then an "or" with the byte value you want to adapt

**Reply by bijlevel** [reply]
Thanks, but I don't understand why 'getBitRangeAsInt' does work and the 'setBitRangeAsInt' doesn't. This way to get and set a bit value looks to me very easy, if it works.... Look here: https://docs.juce.com/master/classBigInteger.html

**Reply by bijlevel** [reply, marked as answer]
Well, I solved it. It was wrong coding. I mistakingly use a '.' instead of ':', so instead of db = flags.setBitRangeAsInt() I had to use flags:setBitRangeAsInt(). So now my actual code looks like this:
`flagSetting = BigInteger(dbSpecial)`

    flagSetting:setBitRangeAsInt(3,1,1)
    dbSpecial = flagSetting:getBitRangeAsInt(0,8)`

So the **set** part changes a bit in the byte to 1 and the **get** part gets the changed value for dbSpecial.

**Reply by dobo365** [reply]
In IT, tere are always different ways to achieve the same result ;-)

**Comment by dnaldoog** [comment]
Hi @bijlevel,

Slightly shorter code is: 

<pre>
flagSetting=BigInteger(dbSpecial)
flagSetting:setBit(3)
dbSpecial=flagSetting:toInteger()
</pre>

**Reply by bijlevel** [reply]
Ah, that looks good! Thanks. Does setBit always set a bit to the opposite it was, so 0 becomes 1 and 1 becomes 0?

**Reply by dnaldoog** [reply]
`setBit` always sets the bit to __1__.

But, it looks like there is a second argument to `setBit` true / false that sets the bit to 1 or 0:

So it looks like you can just do:

<pre>
flagSetting = BigInteger(dbSpecial)
if someValue == 1 then
    flagSetting:setBit(3, true)
else
    flagSetting:setBit(3, false)
end
dbSpecial = flagSetting:toInteger()

</pre>

----

Incidentally, a really useful way of flipping a bit is using __xor__ logic
<pre>
local value=1
value=bit.xor(value,1) -- will return 0

local value=0
value=bit.xor(value,1) -- will return 1
</pre>

**Reply by dnaldoog** [reply]
Or you can do:

<pre>
flagSetting = BigInteger(dbSpecial)
if someValue == 1 then
    flagSetting:setBit(3)
else
    flagSetting:clearBit(3)
end
dbSpecial = flagSetting:toInteger()
</pre>

**Reply by bijlevel** [reply]
Hi John, thanks for expanding my knowledge!
